pennylane/io/qualtran_io.py

```diff
@@ -19,22 +19,589 @@
+
+    qft = _map_to_bloq(qtemps.QFT(wires=range(num_aux_wires)))
+    qft_dag = qft.adjoint()
+
+    sequence = _map_to_bloq(
+        qtemps.ControlledSequence(
+            qtemps.PhaseAdder(k=3, x_wires=range(1, num_x_wires + 1)), control=[0]
+        )
+    )
+    sequence_dag = sequence.adjoint()
+
+    cnot = qt_gates.CNOT()
+
+    mult_resources = {}
+    mult_resources[qft] = 2
+    mult_resources[qft_dag] = 2
+    mult_resources[sequence] = 1
+    mult_resources[sequence_dag] = 1
+    mult_resources[cnot] = min(num_x_wires, num_aux_swap)
+
+    gate_types = defaultdict(int, {})
+    ctrl_spec = CtrlSpec(cvs=[1])
+    for comp_rep in mult_resources:
+        new_rep = comp_rep.controlled(ctrl_spec)
+        if comp_rep == qt_gates.CNOT():
+            new_rep = qt_gates.Toffoli()
+        # cancel out QFTs from consecutive Multipliers
+        if hasattr(comp_rep, "op"):
+            if comp_rep.op.name == "QFT":
+                gate_types[new_rep] = 1
+        elif hasattr(comp_rep, "subbloq"):
+            if comp_rep.subbloq.op.name == "QFT":
+                gate_types[new_rep] = 1
+        else:
+            gate_types[new_rep] = mult_resources[comp_rep] * ((2**num_x_wires) - 1)
+
+    return gate_types
+
+
+@singledispatch
+def _map_to_bloq(op, map_ops=True, custom_mapping=None, **kwargs):
+    """Map PennyLane operators to Qualtran Bloqs. Operators with direct equivalents are directly
+    mapped to their Qualtran equivalent even if ``map_ops`` is set to ``False``. Other operators are
+    given a smart default mapping. When given a ``custom_mapping``, the custom mapping is used."""
+    if not isinstance(op, Operator):
+        return ToBloq(op, map_ops=map_ops, custom_mapping=custom_mapping, **kwargs)
+
+    if custom_mapping is not None:
+        return custom_mapping[op]
+
+    return ToBloq(op, map_ops=map_ops, **kwargs)
+
+
+# pylint: disable=import-outside-toplevel
+@_map_to_bloq.register
+def _(
+    op: qtemps.subroutines.qpe.QuantumPhaseEstimation,
+    map_ops=True,
+    custom_mapping=None,
+    **kwargs,
+):
+    from qualtran.bloqs.phase_estimation import RectangularWindowState
+    from qualtran.bloqs.phase_estimation.text_book_qpe import TextbookQPE
+
+    if not map_ops:
+        return ToBloq(op, **kwargs)
+
+    if custom_mapping is not None:
+        return custom_mapping[op]
+
+    return TextbookQPE(
+        unitary=_map_to_bloq(op.hyperparameters["unitary"]),
+        ctrl_state_prep=RectangularWindowState(len(op.hyperparameters["estimation_wires"])),
+    )
+
+
+def _disable_custom_mapping(func):
+    """Decorator to disable custom mapping and raise error for atomic gates."""
+
+    @wraps(func)
+    def wrapper(op, **kwargs):
+        if kwargs.get("custom_mapping") and op in kwargs.get("custom_mapping", {}):
+            raise ValueError(
+                "Custom mappings are not possible for basic operations. You may map basic operations "
+                "pre or post processing."
```

Comment:
```suggestion
                "Custom mappings are not possible for basic operations. We suggest replacing basic operations "
                "before or after mapping to Qualtran."
```
Also, will it be clear to users what basic operations are? Can we do one of the following?

1. Give examples e.g. X, Y, Z gates
2. Point to where users can see what the basic operations are